File: src/lib/notifications.ts
import { Capacitor } from "@capacitor/core";
import type { Subscription } from "./types";
import { formatDate, relativeDay } from "./format";

/**
 * Native renewal reminders. On a real device (Capacitor) this schedules a local
 * notification a few days before each active subscription renews, with an
 * earlier, louder nudge for manual "must not lapse" bills (insurance / utility).
 *
 * On the web build `Capacitor.isNativePlatform()` is false and every function is
 * a no-op, so the same code ships in the browser without touching the Web
 * Notifications API.
 */

const LEAD_DAYS = 3; // remind this many days before a normal renewal
const LAPSE_LEAD_DAYS = 5; // manual renewals get a bigger head start

const LAPSE_CATEGORIES = new Set(["insurance", "electricity", "gas"]);

/** Stable positive 31-bit integer id from a subscription's string id. */
function notificationId(subId: string): number {
  let h = 0;
  for (let i = 0; i < subId.length; i++) {
    h = (h * 31 + subId.charCodeAt(i)) | 0;
  }
  return (Math.abs(h) % 2_000_000_000) + 1;
}

/** 9am, `offsetDays` before the given ISO renewal date. */
function remindAt(iso: string, offsetDays: number): Date {
  const d = new Date(iso + "T09:00:00");
  d.setDate(d.getDate() - offsetDays);
  return d;
}

/**
 * Reconcile scheduled notifications with the current portfolio: clear the old
 * ones and schedule a fresh reminder for every active, still-upcoming renewal.
 * Safe to call on every change — it fully replaces the previous schedule.
 */
export async function syncRenewalNotifications(
  subs: Subscription[],
): Promise<void> {
  if (!Capacitor.isNativePlatform()) return;

  const { LocalNotifications } = await import("@capacitor/local-notifications");

  const perm = await LocalNotifications.requestPermissions();
  if (perm.display !== "granted") return;

  // Wipe anything we scheduled before so cancellations/edits take effect.
  const pending = await LocalNotifications.getPending();
  if (pending.notifications.length) {
    await LocalNotifications.cancel({
      notifications: pending.notifications.map((n) => ({ id: n.id })),
    });
  }

  const now = Date.now();
  const notifications = [];

  for (const s of subs) {
    if (s.status !== "active") continue;

    const lapseRisk = !s.autoRenew && LAPSE_CATEGORIES.has(s.category);
    const lead = lapseRisk ? LAPSE_LEAD_DAYS : LEAD_DAYS;
    const at = remindAt(s.nextRenewal, lead);
    if (at.getTime() <= now) continue; // never schedule in the past

    notifications.push({
      id: notificationId(s.id),
      title: lapseRisk
        ? `⚠️ ${s.name} — renew to avoid a lapse`
        : `${s.name} renews ${relativeDay(s.nextRenewal)}`,
      body: lapseRisk
        ? `Manual renewal due ${formatDate(s.nextRenewal)}. Pay on time to keep cover and your no-claim bonus.`
        : `Due ${formatDate(s.nextRenewal)}. Open okna to review, switch or cancel if you no longer need it.`,
      schedule: { at },
      smallIcon: "ic_stat_icon_config_sample",
      channelId: "okna-renewals",
    });
  }

  if (notifications.length) {
    await LocalNotifications.schedule({ notifications });
  }
}

File: flutter/okna/lib/services/notifications.dart
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:timezone/data/latest_all.dart' as tzdata;
import 'package:timezone/timezone.dart' as tz;
import '../models/subscription.dart';
import '../logic/format.dart';

/// Native renewal reminders (the Flutter analog of the web app's
/// local-notifications sync). Best-effort — every call is guarded so a
/// notification failure never breaks the app.

final FlutterLocalNotificationsPlugin _plugin = FlutterLocalNotificationsPlugin();
bool _inited = false;

const int _leadDays = 3;
const int _lapseLeadDays = 5;
const Set<CategoryId> _lapseCategories = {
  CategoryId.insurance,
  CategoryId.electricity,
  CategoryId.gas,
};

Future<void> initNotifications() async {
  try {
    tzdata.initializeTimeZones();
    tz.setLocalLocation(tz.getLocation('Asia/Kolkata'));
    const android = AndroidInitializationSettings('@mipmap/ic_launcher');
    await _plugin.initialize(const InitializationSettings(android: android));
    await _plugin
        .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
        ?.requestNotificationsPermission();
    _inited = true;
  } catch (_) {
    _inited = false;
  }
}

int _notificationId(String subId) {
  var h = 0;
  for (final c in subId.codeUnits) {
    h = (h * 31 + c) & 0x7fffffff;
  }
  return (h % 2000000000) + 1;
}

Future<void> syncRenewalNotifications(List<Subscription> subs) async {
  if (!_inited) return;
  try {
    await _plugin.cancelAll();
    const details = NotificationDetails(
      android: AndroidNotificationDetails(
        'okna-renewals',
        'Renewals',
        channelDescription: 'Reminders before your bills and subscriptions renew',
        importance: Importance.high,
        priority: Priority.high,
      ),
    );
    final now = DateTime.now();
    for (final s in subs) {
      if (s.status != SubStatus.active) continue;
      final lapseRisk = !s.autoRenew && _lapseCategories.contains(s.category);
      final lead = lapseRisk ? _lapseLeadDays : _leadDays;
      final at = DateTime.parse('${s.nextRenewal}T09:00:00').subtract(Duration(days: lead));
      if (!at.isAfter(now)) continue;
      final title = lapseRisk
          ? '⚠️ ${s.name} — renew to avoid a lapse'
          : '${s.name} renews ${relativeDay(s.nextRenewal)}';
      final body = lapseRisk
          ? 'Manual renewal due ${formatDate(s.nextRenewal)}. Pay on time to keep cover and your no-claim bonus.'
          : 'Due ${formatDate(s.nextRenewal)}. Open okna to review, switch or cancel if you no longer need it.';
      await _plugin.zonedSchedule(
        _notificationId(s.id),
        title,
        body,
        tz.TZDateTime.from(at, tz.local),
        details,
        androidScheduleMode: AndroidScheduleMode.inexactAllowWhileIdle,
        uiLocalNotificationDateInterpretation: UILocalNotificationDateInterpretation.absoluteTime,
      );
    }
  } catch (_) {
    // best-effort
  }
}
